packages/analytics/shippers/elastic_v3/server/src/server_shipper.ts

```diff
@@ -0,0 +1,279 @@
+  ) {
+    this.setInternalSubscriber();
+    this.checkConnectivity();
+  }
+
+  public extendContext(newContext: EventContext) {
+    if (newContext.cluster_uuid) {
+      this.clusterUuid = newContext.cluster_uuid;
+    }
+    if (newContext.license_id) {
+      this.licenseId = newContext.license_id;
+    }
+  }
+
+  public optIn(isOptedIn: boolean) {
+    this.isOptedIn = isOptedIn;
+
+    if (isOptedIn === false) {
+      this.internalQueue.length = 0;
+    }
+  }
+
+  public reportEvents(events: Event[]) {
+    if (
+      this.isOptedIn === false ||
+      (this.firstTimeOffline && Date.now() - this.firstTimeOffline > 24 * HOUR)
+    ) {
+      return;
+    }
+
+    const freeSpace = 1000 - this.internalQueue.length;
+
+    // As per design, we only want store up-to 1000 events at a time. Drop anything that goes beyond that limit
+    if (freeSpace < events.length) {
+      const droppedEvents = events.slice(-(events.length - freeSpace));
+      events.length = freeSpace;
+      this.reportTelemetryCounters(droppedEvents, {
+        type: TelemetryCounterType.dropped,
+        code: 'queue_full',
+      });
+    }
+
+    this.internalQueue.push(...events);
+  }
+
+  public shutdown() {
+    this.shutdown$.complete();
+  }
+
+  /**
+   * Checks the server has connectivity to the remote endpoint.
+   * The frequency of the connectivity tests will back off, starting with 1 minute, and multiplying by 2
+   * until it reaches 1 hour. Then, it’ll keep the 1h frequency until it reaches 24h without connectivity.
+   * At that point, it clears the queue and stops accepting events in the queue.
+   * The connectivity checks will continue to happen every 1 hour just in case it regains it at any point.
+   * @private
+   */
+  private checkConnectivity() {
+    let backoff = 1 * MINUTE;
+    timer(0, 1 * MINUTE)
+      .pipe(
+        takeWhile(() => this.shutdown$.isStopped === false),
+        filter(() => this.isOptedIn === true && this.firstTimeOffline !== null),
+        exhaustMap(async () => {
+          const { ok } = await fetch(buildUrl(this.initContext.sendTo, this.options.channelName), {
+            method: 'OPTIONS',
+          });
+
+          if (!ok) {
+            throw new Error('Failed to connect to Elastic');
+          }
+
+          this.firstTimeOffline = null;
+          backoff = 1 * MINUTE;
+        }),
+        retryWhen((errors) =>
+          errors.pipe(
+            takeWhile(() => this.shutdown$.isStopped === false),
+            tap(() => {
+              if (!this.firstTimeOffline) {
+                this.firstTimeOffline = Date.now();
+              } else if (Date.now() - this.firstTimeOffline > 24 * HOUR) {
+                this.internalQueue.length = 0;
+              }
+              backoff = backoff * 2;
+              if (backoff > 1 * HOUR) {
+                backoff = 1 * HOUR;
+              }
+            }),
+            delayWhen(() => timer(backoff))
```

Comment:
NIT: I think we can use `delay(backoff)` here instead of `delayWhen(() => timer(backoff))`?

Reply:
the problem with `delay(backoff)` is that it won't apply the new value of `backoff`. We need a lazy evaluation so each retry uses the updated value of backoff.

Reply:
Did not test, but I don't think that's true, as the observable returned by `retryWhen` is reevaluated for each emission, for the value of `backoff` should be up to date?

But it's fine with current implem anyway.